File: Java/Pong.java
package Java;
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.Random;

/**
 * PONG — single-file Java game built with Swing.
 * Controls: Left = W/S   Right = UP/DOWN   Serve = ENTER
 *
 * File structure:
 *   1. Constants        – tweak these to change game feel
 *   2. State variables  – everything that changes at runtime
 *   3. Constructor      – one-time setup
 *   4. update()         – moves objects, checks collisions/scoring
 *   5. paintComponent() – draws the current frame
 *   6. Helpers          – launch, reset, physics utilities
 *   7. main()           – creates the window
 */
public class Pong extends JPanel implements ActionListener, KeyListener {

    // ── 1. CONSTANTS ─────────────────────────────────────────
    static final int W = 900, H = 600;
    static final int PADDLE_W = 14, PADDLE_H = 90, INSET = 30;
    static final int BALL_SIZE = 16;
    static final int PADDLE_SPEED = 6;
    static final int WIN_SCORE = 7;
    static final double INIT_SPEED = 5, SPEED_INC = 0.4, MAX_SPEED = 14;

    // ── 2. STATE ──────────────────────────────────────────────
    int lpY, rpY;                          // left/right paddle Y
    double bX, bY, bVX, bVY;              // ball position & velocity
    int lScore, rScore;
    boolean lUp, lDown, rUp, rDown;        // held keys
    String state = "waiting";             // waiting | playing | gameover
    Random rng = new Random();

    // ── 3. CONSTRUCTOR ────────────────────────────────────────
    public Pong() {
        setPreferredSize(new Dimension(W, H));
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);
        resetPositions();
        new Timer(16, this).start();       // ~60 FPS game loop
    }

    // ── 4. UPDATE (game logic, runs every frame) ──────────────
    @Override
    public void actionPerformed(ActionEvent e) { update(); repaint(); }

    void update() {
        // Move paddles
        if (lUp   && lpY > 0)            lpY -= PADDLE_SPEED;
        if (lDown && lpY < H - PADDLE_H) lpY += PADDLE_SPEED;
        if (rUp   && rpY > 0)            rpY -= PADDLE_SPEED;
        if (rDown && rpY < H - PADDLE_H) rpY += PADDLE_SPEED;

        if (!state.equals("playing")) return;

        bX += bVX;
        bY += bVY;

        // Top / bottom wall bounce
        if (bY - BALL_SIZE / 2.0 <= 0)  { bY = BALL_SIZE / 2.0;      bVY =  Math.abs(bVY); }
        if (bY + BALL_SIZE / 2.0 >= H)  { bY = H - BALL_SIZE / 2.0;  bVY = -Math.abs(bVY); }

        // Left paddle collision
        double lFace = INSET + PADDLE_W;
        if (bVX < 0 && bX - BALL_SIZE/2.0 <= lFace && bX + BALL_SIZE/2.0 >= INSET
                && bY >= lpY && bY <= lpY + PADDLE_H) {
            bX = lFace + BALL_SIZE / 2.0;
            bVX = Math.abs(bVX);
            bVY += english(bY, lpY);
            speedUp();
        }

        // Right paddle collision
        double rFace = W - INSET - PADDLE_W;
        if (bVX > 0 && bX + BALL_SIZE/2.0 >= rFace && bX - BALL_SIZE/2.0 <= W - INSET
                && bY >= rpY && bY <= rpY + PADDLE_H) {
            bX = rFace - BALL_SIZE / 2.0;
            bVX = -Math.abs(bVX);
            bVY += english(bY, rpY);
            speedUp();
        }

        // Scoring
        if (bX + BALL_SIZE / 2.0 < 0)  { rScore++; nextPoint(); }
        if (bX - BALL_SIZE / 2.0 > W)  { lScore++; nextPoint(); }
    }

    // ── 5. RENDERING ─────────────────────────────────────────
    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2 = (Graphics2D) g;
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        // Center dashed line
        g2.setColor(new Color(255, 255, 255, 50));
        g2.setStroke(new BasicStroke(2, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL, 0, new float[]{12, 12}, 0));
        g2.drawLine(W / 2, 0, W / 2, H);
        g2.setStroke(new BasicStroke(1));

        // Scores
        g2.setColor(Color.WHITE);
        g2.setFont(new Font("Monospaced", Font.BOLD, 64));
        FontMetrics fm = g2.getFontMetrics();
        g2.drawString(String.valueOf(lScore), W/2 - fm.stringWidth(String.valueOf(lScore)) - 40, 80);
        g2.drawString(String.valueOf(rScore), W/2 + 40, 80);

        // Paddles & ball
        g2.fillRoundRect(INSET, lpY, PADDLE_W, PADDLE_H, 6, 6);
        g2.fillRoundRect(W - INSET - PADDLE_W, rpY, PADDLE_W, PADDLE_H, 6, 6);
        g2.fillOval((int)(bX - BALL_SIZE/2.0), (int)(bY - BALL_SIZE/2.0), BALL_SIZE, BALL_SIZE);

        // Overlay messages
        if (state.equals("waiting"))
            centered(g2, "Press ENTER to serve", H/2 + 30, 22, new Color(200, 200, 200));
        if (state.equals("gameover")) {
            centered(g2, (lScore >= WIN_SCORE ? "LEFT" : "RIGHT") + " PLAYER WINS!", H/2 - 20, 40, Color.YELLOW);
            centered(g2, "Press ENTER to play again", H/2 + 35, 22, new Color(200, 200, 200));
        }

        // Control hints
        g2.setColor(new Color(100, 100, 100));
        g2.setFont(new Font("Monospaced", Font.PLAIN, 13));
        g2.drawString("Left: W/S", 20, H - 12);
        String hint = "Right: \u2191/\u2193";
        g2.drawString(hint, W - g2.getFontMetrics().stringWidth(hint) - 20, H - 12);
    }

    // ── 6. HELPERS ────────────────────────────────────────────

    void launchBall() {
        double angle = Math.toRadians(rng.nextInt(41) - 20);
        int dir = rng.nextBoolean() ? 1 : -1;
        bVX = dir * INIT_SPEED * Math.cos(angle);
        bVY =       INIT_SPEED * Math.sin(angle);
    }

    void resetPositions() {
        bX = W / 2.0;  bY = H / 2.0;  bVX = 0;  bVY = 0;
        lpY = rpY = (H - PADDLE_H) / 2;
    }

    void nextPoint() {
        resetPositions();
        state = (lScore >= WIN_SCORE || rScore >= WIN_SCORE) ? "gameover" : "waiting";
    }

    // How far from center the ball hit the paddle → steeper angle at edges
    double english(double ballCY, double paddleTopY) {
        return ((ballCY - paddleTopY) / PADDLE_H - 0.5) * 4.0;
    }

    // Increase speed after each paddle hit, capped at MAX_SPEED
    void speedUp() {
        double spd = Math.sqrt(bVX*bVX + bVY*bVY);
        double scale = Math.min(spd + SPEED_INC, MAX_SPEED) / spd;
        bVX *= scale;  bVY *= scale;
    }

    void centered(Graphics2D g2, String text, int y, int size, Color c) {
        g2.setColor(c);
        g2.setFont(new Font("Monospaced", Font.BOLD, size));
        g2.drawString(text, (W - g2.getFontMetrics().stringWidth(text)) / 2, y);
    }

    // ── 7. INPUT ─────────────────────────────────────────────
    @Override
    public void keyPressed(KeyEvent e) {
        switch (e.getKeyCode()) {
            case KeyEvent.VK_W     -> lUp    = true;
            case KeyEvent.VK_S     -> lDown  = true;
            case KeyEvent.VK_UP    -> rUp    = true;
            case KeyEvent.VK_DOWN  -> rDown  = true;
            case KeyEvent.VK_ENTER -> {
                if (state.equals("waiting"))  { launchBall(); state = "playing"; }
                if (state.equals("gameover")) { lScore = 0; rScore = 0; resetPositions(); state = "waiting"; }
            }
        }
    }
    @Override
    public void keyReleased(KeyEvent e) {
        switch (e.getKeyCode()) {
            case KeyEvent.VK_W    -> lUp    = false;
            case KeyEvent.VK_S    -> lDown  = false;
            case KeyEvent.VK_UP   -> rUp    = false;
            case KeyEvent.VK_DOWN -> rDown  = false;
        }
    }
    @Override public void keyTyped(KeyEvent e) {}

    // ── 8. MAIN ───────────────────────────────────────────────
    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            JFrame f = new JFrame("Pong");
            f.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
            f.setResizable(false);
            Pong game = new Pong();
            f.add(game);
            f.pack();
            f.setLocationRelativeTo(null);
            f.setVisible(true);
            game.requestFocusInWindow();
        });
    }
}

File: Python/pong.py
import pygame
import sys
import math
import random

"""
PONG — single-file Python game built with pygame.
Controls: Left = W/S   Right = UP/DOWN   Serve = ENTER

File structure:
  1. Constants       – tweak these to change game feel
  2. State           – everything that changes at runtime
  3. reset()         – re-centers ball and paddles
  4. launch_ball()   – randomizes starting velocity
  5. update()        – moves objects, checks collisions/scoring
  6. draw()          – renders the current frame
  7. Main loop       – event handling + game loop
"""

pygame.init()

# ── 1. CONSTANTS
W, H = 900, 600
PADDLE_W = 14
PADDLE_H = 90
INSET = 30
BALL_SIZE = 16
PADDLE_SPEED = 6
WIN_SCORE = 7
INIT_SPEED = 5.0
SPEED_INC = 0.4
MAX_SPEED = 14.0
FPS = 60

WHITE = (255, 255, 255)
BLACK = (0,   0,   0)
YELLOW = (255, 255, 0)
DIM = (200, 200, 200)
GREY = (100, 100, 100)

screen = pygame.display.set_mode((W, H))
pygame.display.set_caption("Pong")
clock = pygame.time.Clock()

font_big = pygame.font.SysFont("Courier", 64, bold=True)
font_mid = pygame.font.SysFont("Courier", 40, bold=True)
font_small = pygame.font.SysFont("Courier", 22, bold=True)
font_hint = pygame.font.SysFont("Courier", 13)


# ── 2. STATE────
lp_y = rp_y = (H - PADDLE_H) // 2  # paddle Y positions
bx = by = 0.0                       # ball center
bvx = bvy = 0.0                     # ball velocity
l_score = r_score = 0
state = "waiting"                   # waiting | playing | gameover
rng = random.Random()


# ── 3. RESET────
def reset():
    global lp_y, rp_y, bx, by, bvx, bvy
    bx, by = W / 2, H / 2
    bvx = bvy = 0
    lp_y = rp_y = (H - PADDLE_H) // 2


# ── 4. LAUNCH BALL
def launch_ball():
    global bvx, bvy
    angle = math.radians(rng.randint(-20, 20))
    d = 1 if rng.random() > 0.5 else -1
    bvx = d * INIT_SPEED * math.cos(angle)
    bvy = INIT_SPEED * math.sin(angle)


# ── 5. UPDATE───
def update(keys):
    global lp_y, rp_y, bx, by, bvx, bvy
    global l_score, r_score, state

    if keys[pygame.K_w] and lp_y > 0:
        lp_y -= PADDLE_SPEED
    if keys[pygame.K_s] and lp_y < H - PADDLE_H:
        lp_y += PADDLE_SPEED
    if keys[pygame.K_UP] and rp_y > 0:
        rp_y -= PADDLE_SPEED
    if keys[pygame.K_DOWN] and rp_y < H - PADDLE_H:
        rp_y += PADDLE_SPEED

    if state != "playing":
        return

    bx += bvx
    by += bvy

    # Wall bounce
    if by - BALL_SIZE / 2 <= 0:
        by = BALL_SIZE / 2
        bvy = abs(bvy)
    if by + BALL_SIZE / 2 >= H:
        by = H - BALL_SIZE / 2
        bvy = -abs(bvy)

    # Left paddle collision
    l_face = INSET + PADDLE_W
    if (bvx < 0
            and bx - BALL_SIZE / 2 <= l_face
            and bx + BALL_SIZE / 2 >= INSET
            and lp_y <= by <= lp_y + PADDLE_H):
        bx = l_face + BALL_SIZE / 2
        bvx = abs(bvx)
        bvy += calcDeflectionAngle(by, lp_y)
        speed_up()

    # Right paddle collision
    r_face = W - INSET - PADDLE_W
    if (bvx > 0
            and bx + BALL_SIZE / 2 >= r_face
            and bx - BALL_SIZE / 2 <= W - INSET
            and rp_y <= by <= rp_y + PADDLE_H):
        bx = r_face - BALL_SIZE / 2
        bvx = -abs(bvx)
        bvy += calcDeflectionAngle(by, rp_y)
        speed_up()

    # Scoring
    if bx + BALL_SIZE / 2 < 0:
        r_score += 1
        next_point()
    if bx - BALL_SIZE / 2 > W:
        l_score += 1
        next_point()


def calcDeflectionAngle(ball_cy, paddle_top):
    return ((ball_cy - paddle_top) / PADDLE_H - 0.5) * 4.0


def speed_up():
    global bvx, bvy
    spd = math.sqrt(bvx**2 + bvy**2)
    scale = min(spd + SPEED_INC, MAX_SPEED) / spd
    bvx *= scale
    bvy *= scale


def next_point():
    global state
    reset()
    won = l_score >= WIN_SCORE or r_score >= WIN_SCORE
    state = "gameover" if won else "waiting"


# ── 6. DRAW─────
def centered(text, y, font, color):
    s = font.render(text, True, color)
    screen.blit(s, ((W - s.get_width()) // 2, y))


def draw():
    screen.fill(BLACK)

    # Dashed center line
    for yy in range(0, H, 24):
        pygame.draw.rect(screen, GREY, (W // 2 - 1, yy, 2, 12))

    # Scores
    ls = font_big.render(str(l_score), True, WHITE)
    rs = font_big.render(str(r_score), True, WHITE)
    screen.blit(ls, (W // 2 - ls.get_width() - 40, 20))
    screen.blit(rs, (W // 2 + 40, 20))

    # Paddles & ball
    pygame.draw.rect(
        screen, WHITE, (INSET, lp_y, PADDLE_W, PADDLE_H), border_radius=4
    )
    pygame.draw.rect(
        screen, WHITE,
        (W - INSET - PADDLE_W, rp_y, PADDLE_W, PADDLE_H), border_radius=4
    )
    pygame.draw.circle(screen, WHITE, (int(bx), int(by)), BALL_SIZE // 2)

    # Overlay messages
    if state == "waiting":
        centered("Press ENTER to serve", H // 2 + 30, font_small, DIM)
    elif state == "gameover":
        winner = "LEFT" if l_score >= WIN_SCORE else "RIGHT"
        centered(f"{winner} PLAYER WINS!", H // 2 - 20, font_mid, YELLOW)
        centered("Press ENTER to play again", H // 2 + 35, font_small, DIM)

    # Control hints
    screen.blit(font_hint.render("Left: W/S", True, GREY), (20, H - 20))
    rh = font_hint.render("Right: \u2191/\u2193", True, GREY)
    screen.blit(rh, (W - rh.get_width() - 20, H - 20))

    pygame.display.flip()


# ── 7. MAIN LOOP
reset()
while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        if event.type == pygame.KEYDOWN and event.key == pygame.K_RETURN:
            if state == "waiting":
                launch_ball()
                state = "playing"
            elif state == "gameover":
                l_score = r_score = 0
                reset()
                state = "waiting"

    update(pygame.key.get_pressed())
    draw()
    clock.tick(FPS)
